src/effects/native/gatereffect.cpp

```diff
@@ -0,0 +1,220 @@
+        if(groupFeatures.has_beat_length_sec && groupFeatures.has_beat_fraction) {
+            timePosition = groupFeatures.beat_fraction*groupFeatures.beat_length_sec*sampleRate;
+            beatPeriod   = groupFeatures.beat_length_sec*sampleRate;
+        } else {
+            timePosition = 0;   
+            beatPeriod   = sampleRate; // 1 second
+        }
+    }
+
+    // Adjust trigger period
+    unsigned int triggerPeriod;
+    double period = 1-rate;
+    if (groupFeatures.has_beat_length_sec) {
+        // triggerPeriod is a number of beats
+        if (m_pQuantizeParameter->toBool()) {
+            // Divide a beat by the divider factor
+            period = 1.0/divider;
+            
+            if (m_pTripletParameter->toBool()) {
+                period /= 3.0;
+            }
+        } else if (period < 1 / 8.0) {
+            period = 1 / 8.0;
+        }
+        triggerPeriod = period * groupFeatures.beat_length_sec * sampleRate;
+    } else {
+        // triggerPeriod is a number of seconds
+        period = std::max(period, 1 / 8.0);
+        triggerPeriod = period * sampleRate;
+    }
+
+    // triggerTime : offset for the gate trigger
+    unsigned int triggerTime = m_pInvertParameter->toBool() ? triggerPeriod/2 : 0;
+    // holdTime : number of samples spent in the Hold state
+    unsigned int holdTime = std::max(0, (int)(triggerPeriod*shape - maxGain/attackInc - maxGain/releaseInc));
+    
+
+    for(unsigned int i = 0; i < numSamples; i+=2) {
+         if (timePosition % triggerPeriod == triggerTime) {
+            state = GaterGroupState::ATTACK;
+            if(timePosition % beatPeriod == 1)
+                timePosition = 0;
+        }
+        timePosition++;
+
+        // Gate state machine : Idle->Attack->Hold->Release->Idle
+        switch(state) {
+            // Idle : Gain = 0
+            case GaterGroupState::IDLE:
+                gain = 0;
+                break;
+
+            // Attack : Increase gain up to maxGain, then Hold
+            case GaterGroupState::ATTACK:
+                gain += attackInc;
+                if(gain >= maxGain) {
+                    state = GaterGroupState::HOLD;
+                    holdCounter = holdTime;
+                }
+                break;
+
+            // Hold : Hold maxGain for holdCounter samples
+            case GaterGroupState::HOLD:
+                gain = maxGain;
+                if(holdCounter == 0)
+                    state = GaterGroupState::RELEASE;  
+
+                holdCounter--;
+                break;
+
+            // Release : Decrease gain to 0, then Idle
+            case GaterGroupState::RELEASE:
+                gain -= releaseInc;
+                if(gain <= 0)
+                    state = GaterGroupState::IDLE;
+                break;
+        }
+
+        pOutput[i]   = gain*pInput[i];
+        pOutput[i+1] = gain*pInput[i+1];
+    }
+
+    // Write back channel state
+    pState->timePosition[handle] = timePosition;
```

Comment:
There is no need to add [handle] because there is already passed the handle's pState.